### PR TITLE
Enable Compute Follows Data in Cython in fft and linalg

### PR DESCRIPTION
**Pull request by LukichevaPolina**
The PR is intended to apply Compute Follows Data approach for all supported DPNP functions of `fft` and `linalg` libraries.
- The implemented changes include an adaptation in python, cython and backend parts associated with `fft` and `linalg` functionality.
- Additional tests are introduced in `test_sycl_queue.py` to secure Compute Follows Data support for `fft` and `linalg` in DPNP. The new tests assume running checks for equality of underlying `backend`, `sycl_device`, `sycl_context` and properties of argument sycl queues at the end of verifying sequence.
- The tests from `test_fft.py`, `test_linalg.py` and `test_sycl_queue.py` (covering `fft` and `linalg` functionality) are added to a validation jobs running as github actions per each commit in DPNP.

### CODE REVIEW

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 1]
```diff
@@ -655,23 +665,26 @@ DPCTLSyclEventRef dpnp_rng_hypergeometric_c(DPCTLSyclQueueRef q_ref,
...
     else if (l == m)
     {
         _DataType* fill_value = reinterpret_cast<_DataType*>(sycl::malloc_shared(sizeof(_DataType), q));
         fill_value[0] = static_cast<_DataType>(s);
-        dpnp_initval_c<_DataType>(result, fill_value, size);
-        sycl::free(fill_value, q);
+
+        event_ref = dpnp_initval_c<_DataType>(q_ref, result, fill_value, size, dep_event_vec_ref);
+        DPCTLEvent_Wait(event_ref);
+        dpnp_memory_free_c(fill_value);
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:677]: Please pass `q_ref` as argument.

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 2]
```diff
@@ -864,19 +879,21 @@ DPCTLSyclEventRef dpnp_rng_lognormal_c(DPCTLSyclQueueRef q_ref,
     {
         _DataType* fill_value = reinterpret_cast<_DataType*>(sycl::malloc_shared(sizeof(_DataType), q));
         fill_value[0] = static_cast<_DataType>(std::exp(mean + (stddev * stddev) / 2));
-        dpnp_initval_c<_DataType>(result, fill_value, size);
-        sycl::free(fill_value, q);
+
+        event_ref = dpnp_initval_c<_DataType>(q_ref, result, fill_value, size, dep_event_vec_ref);
+        DPCTLEvent_Wait(event_ref);
+        dpnp_memory_free_c(fill_value);
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:885]: Please pass `q_ref` as argument.

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 3]
```diff
@@ -2760,96 +2786,190 @@ void func_map_init_random(func_map_t& fmap)
...
     fmap[DPNPFuncName::DPNP_FN_RNG_SHUFFLE][eft_LNG][eft_LNG] = {eft_LNG, (void*)dpnp_rng_shuffle_default_c<int64_t>};
 
+    fmap[DPNPFuncName::DPNP_FN_RNG_SHUFFLE_EXT][eft_DBL][eft_DBL] = {eft_DBL, (void*)dpnp_rng_shuffle_ext_c<double>};
+    fmap[DPNPFuncName::DPNP_FN_RNG_SHUFFLE_EXT][eft_FLT][eft_FLT] = {eft_FLT, (void*)dpnp_rng_shuffle_ext_c<float>};
+    fmap[DPNPFuncName::DPNP_FN_RNG_SHUFFLE_EXT][eft_INT][eft_INT] = {eft_INT, (void*)dpnp_rng_shuffle_ext_c<int32_t>};
+    fmap[DPNPFuncName::DPNP_FN_RNG_SHUFFLE_EXT][eft_LNG][eft_LNG] = {eft_LNG, (void*)dpnp_rng_shuffle_ext_c<int64_t>};
+
     fmap[DPNPFuncName::DPNP_FN_RNG_SRAND][eft_DBL][eft_DBL] = {eft_DBL, (void*)dpnp_rng_srand_c};
 
+    // fmap[DPNPFuncName::DPNP_FN_RNG_SRAND_EXT][eft_DBL][eft_DBL] = {eft_DBL, (void*)dpnp_rng_srand_c};
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:2912]: The line is not needed.

File: dpnp/backend/kernels/dpnp_krnl_random.cpp [hunk 4]
```diff
@@ -2760,96 +2786,190 @@ void func_map_init_random(func_map_t& fmap)
...
         eft_DBL, (void*)dpnp_rng_standard_normal_default_c<double>};
+
+    fmap[DPNPFuncName::DPNP_FN_RNG_STANDARD_NORMAL_EXT][eft_DBL][eft_DBL] = {
+        eft_DBL, (void*)dpnp_rng_standard_normal_ext_c<double>};
+
     fmap[DPNPFuncName::DPNP_FN_RNG_STANDARD_T][eft_DBL][eft_DBL] = {
         eft_DBL, (void*)dpnp_rng_standard_t_default_c<double>};
 
+    fmap[DPNPFuncName::DPNP_FN_RNG_STANDARD_T_EXT][eft_DBL][eft_DBL] = {
+        eft_DBL, (void*)dpnp_rng_standard_normal_ext_c<double>};
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_random.cpp:2942]:
```suggestion
        eft_DBL, (void*)dpnp_rng_standard_t_ext_c<double>};
```

File: dpnp/linalg/dpnp_algo_linalg.pyx [hunk 5]
```diff
@@ -311,15 +447,48 @@ cpdef tuple dpnp_qr(utils.dpnp_descriptor x1, str mode):
...
-    func(x1.get_data(), res_q.get_data(), res_r.get_data(), tau.get_data(), size_m, size_n)
+    cdef c_dpctl.DPCTLSyclEventRef event_ref = func(q_ref,
+                                                    x1.get_data(),
+                                                    res_q.get_data(),
+                                                    res_r.get_data(),
+                                                    tau.get_data(),
+                                                    size_m,
+                                                    size_n,
+                                                    NULL)
+                                                     # dep_events_ref
```
Comment (densmirn) [inline dpnp/linalg/dpnp_algo_linalg.pyx:488]:
```suggestion
                                                    NULL)  # dep_events_ref
```

File: dpnp/random/dpnp_algo_random.pyx [hunk 6]
```diff
@@ -540,16 +779,23 @@ cpdef utils.dpnp_descriptor dpnp_rng_multivariate_normal(numpy.ndarray mean, num
...
+    cdef c_dpctl.DPCTLSyclQueueRef q_ref = q.get_queue_ref()
 
     func = <fptr_dpnp_rng_multivariate_normal_c_1out_t > kernel_data.ptr
     # call FPTR function
-    func(result.get_data(), dimen, mean_vector, mean_vector_size, cov_vector, cov_vector_size, result.size)
+    func(result.get_data(), mean_size, mean.get_data(), mean_size, cov.get_data(), cov_size, result.size)
 
     return result
-
+
```
Comment (densmirn) [inline dpnp/random/dpnp_algo_random.pyx:798]:
```suggestion

```

File: dpnp/random/dpnp_algo_random.pyx [hunk 7]
```diff
@@ -778,15 +1089,20 @@ cpdef utils.dpnp_descriptor dpnp_rng_random(dims):
...
 
+    result_sycl_queue = result.get_array().sycl_queue
+
+    cdef c_dpctl.SyclQueue q = <c_dpctl.SyclQueue> result_sycl_queue
+    cdef c_dpctl.DPCTLSyclQueueRef q_ref = q.get_queue_ref()
+
     cdef fptr_dpnp_rng_uniform_c_1out_t func = <fptr_dpnp_rng_uniform_c_1out_t > kernel_data.ptr
     # call FPTR function
-    func(result.get_data(), low, high, result.size)
+    cdef c_dpctl.DPCTLSyclEventRef event_ref = func(q_ref, result.get_data(), low, high, result.size, NULL)
```
Comment (densmirn) [inline dpnp/random/dpnp_algo_random.pyx:1105]: Didn't you forget to wait for result?

**Review by densmirn — APPROVED** [review]
We need to wait for successful testing in CI.

**Review by densmirn — APPROVED** [review]

**Comment by antonwolfy** [comment]
The is no way to trigger tasks from internal CI, because the PR is too old and not owned by me. It is going to be secured after merge manually if no objections.

**Comment by oleksandr-pavlyk** [comment]
I used artifact from public CI to test the functionality and used `onetrace` utility to verify that a single execution queue was used when executing each of modified functions. 

Approving

**Pull request merged**